File: backend/src/baserow/contrib/database/formula/parser/replace_field_by_id_with_field.py
from io import StringIO
from typing import Dict

from antlr4 import Token

from baserow.contrib.database.formula.parser.generated.BaserowFormulaLexer import (
    BaserowFormulaLexer,
)
from baserow.contrib.database.formula.parser.parser import get_token_stream_for_formula


# Translated directly from replaceFieldByIdWithField.js please keep in sync
# if changes made.
def replace_field_by_id_with_field(
    formula: str, field_id_to_name: Dict[int, str]
) -> str:
    """
    Given a baserow formula transforms any field_by_id(X) references to field(NAME)
    when there is a mapping in field_id_to_name of {X: NAME}. If no mapping is found
    then the field_by_id is left untransformed. Preserves whitespace and comments in
    the returned string.

    :param formula: A string possibly in the Baserow Formula language.
    :param field_id_to_name: A mapping of field ids to field names to replace in the
        formula.
    :return: A formula string with replacements done.
    """

    stream = get_token_stream_for_formula(formula)

    searching_for_open_paren = False
    searching_for_close_paren = False
    searching_for_integer_literal = False

    with StringIO() as buf:
        for i in range(0, len(stream.tokens)):
            t = stream.tokens[i]
            output = t.text

            is_normal_token = t.channel == 0
            if is_normal_token:
                if searching_for_integer_literal:
                    searching_for_integer_literal = False
                    searching_for_close_paren = True
                    if t.type == BaserowFormulaLexer.INTEGER_LITERAL:
                        int_literal = int(t.text)
                        field_name = field_id_to_name[int_literal]
                        escaped_field_name = field_name.replace("'", "\\'")
                        output = f"'{escaped_field_name}'"
                    else:
                        return formula
                elif searching_for_open_paren:
                    searching_for_open_paren = False
                    if t.type == BaserowFormulaLexer.OPEN_PAREN:
                        searching_for_integer_literal = True
                    else:
                        return formula
                elif searching_for_close_paren:
                    searching_for_close_paren = False
                    if t.type != BaserowFormulaLexer.CLOSE_PAREN:
                        return formula
                elif t.type == BaserowFormulaLexer.FIELDBYID:
                    looked_ahead_id = _lookahead_to_id(i + 1, stream)
                    if (
                        looked_ahead_id is not None
                        and looked_ahead_id in field_id_to_name
                    ):
                        output = "field"
                        searching_for_open_paren = True

            if t.type == Token.EOF:
                break

            buf.write(output)
        return buf.getvalue()


def _lookahead_to_id(start, stream):
    searching_for_int_literal = False
    for i in range(start, len(stream.tokens)):
        t = stream.tokens[i]
        is_normal_token = t.channel == 0
        if is_normal_token:
            if searching_for_int_literal:
                if t.type == BaserowFormulaLexer.INTEGER_LITERAL:
                    return int(t.text)
                else:
                    return None
            elif t.type == BaserowFormulaLexer.OPEN_PAREN:
                searching_for_int_literal = True
            else:
                return None
        if t.type == Token.EOF:
            return None
    return None

File: web-frontend/modules/database/formula/parser/replaceFieldByIdWithField.js
import { BaserowFormulaLexer } from '@baserow/modules/database/formula/parser/generated/BaserowFormulaLexer'
import { getTokenStreamForFormula } from '@baserow/modules/database/formula/parser/parser'

/**
 * Given a map of field id to field name replaces all field_by_id references to
 * with field references. Does so whist preserving any whitespace or
 * comments. If a reference to an unknown field_by_id is made it will be left as it is.
 *
 * This algorithm is duplicated in the backend in replace_field_by_id_with_field.py
 * please sync across any changes.
 *
 * @param formula The raw string to tokenize and transform.
 * @param fieldIdToName The map of field ids to names.
 * @returns string False if the formula is not
 *    syntactically correct, otherwise the new updated formula string.
 */
export function replaceFieldByIdWithField(formula, fieldIdToName) {
  const stream = getTokenStreamForFormula(formula)

  let searchingForOpenParen = false
  let searchingForCloseParen = false
  let searchingForIntegerLiteral = false

  let newFormula = ''
  for (let i = 0; i < stream.tokens.length; i++) {
    const token = stream.tokens[i]
    let output = token.text

    const isNormalToken = token.channel === 0
    if (isNormalToken) {
      if (searchingForIntegerLiteral) {
        searchingForIntegerLiteral = false
        searchingForCloseParen = true
        if (token.type === BaserowFormulaLexer.INTEGER_LITERAL) {
          output = `'${fieldIdToName[output].replace("'", "\\'")}'`
        } else {
          // The only valid normal token is an int, we've encountered a different
          // token and hence the input string is invalid and so we'll just return it
          // untouched.
          return formula
        }
      } else if (searchingForOpenParen) {
        searchingForOpenParen = false
        if (token.type === BaserowFormulaLexer.OPEN_PAREN) {
          searchingForIntegerLiteral = true
        } else {
          return formula
        }
      } else if (searchingForCloseParen) {
        searchingForCloseParen = false
        if (token.type !== BaserowFormulaLexer.CLOSE_PAREN) {
          return formula
        }
      } else if (token.type === BaserowFormulaLexer.FIELDBYID) {
        const futureIntLiteral = _lookaheadAndFindFieldByIdIntLiteral(
          i + 1,
          stream
        )
        if (
          futureIntLiteral !== false &&
          fieldIdToName[futureIntLiteral] !== undefined
        ) {
          // Only replace field_by_id with field if we know we will find a proper int
          // referenced we know about. Otherwise we want to leave the field_by_id
          // untouched.
          searchingForOpenParen = true
          output = 'field'
        }
      }
    }

    if (token.type === BaserowFormulaLexer.EOF) {
      break
    }

    newFormula += output
  }
  return newFormula
}

function _lookaheadAndFindFieldByIdIntLiteral(start, stream) {
  let searchingForIntegerLiteral = false
  for (let i = start; i < stream.tokens.length; i++) {
    const token = stream.tokens[i]
    const isNormalToken = token.channel === 0

    if (isNormalToken) {
      if (searchingForIntegerLiteral) {
        if (token.type === BaserowFormulaLexer.INTEGER_LITERAL) {
          return parseInt(token.text)
        } else {
          return false
        }
      } else if (token.type === BaserowFormulaLexer.OPEN_PAREN) {
        searchingForIntegerLiteral = true
      } else {
        return false
      }
    }
    if (token.type === BaserowFormulaLexer.EOF) {
      return false
    }
  }
  return false
}
